File: crates/pact-bindings-core/src/capability.rs
use pact_core::{validate_delegation_chain, CapabilityToken, Error as CoreError};
use serde::{Deserialize, Serialize};

use crate::Result;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CapabilityTimeStatus {
    Valid,
    NotYetValid,
    Expired,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct CapabilityVerification {
    pub signature_valid: bool,
    pub delegation_chain_valid: bool,
    pub time_valid: bool,
    pub time_status: CapabilityTimeStatus,
}

pub fn parse_capability_json(input: &str) -> Result<CapabilityToken> {
    Ok(serde_json::from_str(input)?)
}

pub fn capability_body_canonical_json(capability: &CapabilityToken) -> Result<String> {
    pact_core::canonical_json_string(&capability.body()).map_err(Into::into)
}

pub fn verify_capability(
    capability: &CapabilityToken,
    now: u64,
    max_delegation_depth: Option<u32>,
) -> Result<CapabilityVerification> {
    let time_status = match capability.validate_time(now) {
        Ok(()) => CapabilityTimeStatus::Valid,
        Err(CoreError::CapabilityNotYetValid { .. }) => CapabilityTimeStatus::NotYetValid,
        Err(CoreError::CapabilityExpired { .. }) => CapabilityTimeStatus::Expired,
        Err(error) => return Err(error.into()),
    };

    Ok(CapabilityVerification {
        signature_valid: capability.verify_signature()?,
        delegation_chain_valid: validate_delegation_chain(
            &capability.delegation_chain,
            max_delegation_depth,
        )
        .is_ok(),
        time_valid: matches!(time_status, CapabilityTimeStatus::Valid),
        time_status,
    })
}

pub fn verify_capability_json(
    input: &str,
    now: u64,
    max_delegation_depth: Option<u32>,
) -> Result<CapabilityVerification> {
    let capability = parse_capability_json(input)?;
    verify_capability(&capability, now, max_delegation_depth)
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used)]
mod tests {
    use super::{verify_capability, CapabilityTimeStatus};
    use pact_core::{
        CapabilityToken, CapabilityTokenBody, Constraint, Keypair, Operation, PactScope, ToolGrant,
    };

    fn sample_scope() -> PactScope {
        PactScope {
            grants: vec![ToolGrant {
                server_id: "srv-files".to_string(),
                tool_name: "file_read".to_string(),
                operations: vec![Operation::Invoke, Operation::ReadResult],
                constraints: vec![Constraint::PathPrefix("/workspace/".to_string())],
                max_invocations: Some(3),
                max_cost_per_invocation: None,
                max_total_cost: None,
            }],
            resource_grants: vec![],
            prompt_grants: vec![],
        }
    }

    fn sample_capability() -> CapabilityToken {
        let issuer = Keypair::from_seed(&[11u8; 32]);
        let subject = Keypair::from_seed(&[12u8; 32]);
        CapabilityToken::sign(
            CapabilityTokenBody {
                id: "cap-bindings-valid".to_string(),
                issuer: issuer.public_key(),
                subject: subject.public_key(),
                scope: sample_scope(),
                issued_at: 1710000200,
                expires_at: 1710000800,
                delegation_chain: vec![],
            },
            &issuer,
        )
        .unwrap()
    }

    #[test]
    fn verify_valid_capability() {
        let capability = sample_capability();
        let verification = verify_capability(&capability, 1710000400, None).unwrap();
        assert_eq!(
            verification,
            super::CapabilityVerification {
                signature_valid: true,
                delegation_chain_valid: true,
                time_valid: true,
                time_status: CapabilityTimeStatus::Valid,
            }
        );
    }
}

File: packages/sdk/arc-go/invariants/capability.go
package invariants

type CapabilityTimeStatus string

const (
	CapabilityTimeStatusValid       CapabilityTimeStatus = "valid"
	CapabilityTimeStatusNotYetValid CapabilityTimeStatus = "not_yet_valid"
	CapabilityTimeStatusExpired     CapabilityTimeStatus = "expired"
)

type CapabilityVerification struct {
	DelegationChainValid bool                 `json:"delegation_chain_valid"`
	SignatureValid       bool                 `json:"signature_valid"`
	TimeStatus           CapabilityTimeStatus `json:"time_status"`
	TimeValid            bool                 `json:"time_valid"`
}

func ParseCapabilityJSON(input string) (map[string]any, error) {
	value, err := ParseJSONText(input)
	if err != nil {
		return nil, err
	}
	capability, ok := value.(map[string]any)
	if !ok {
		return nil, newInvariantError("json", "capability must be a JSON object")
	}
	return capability, nil
}

func CapabilityBodyCanonicalJSON(capability map[string]any) (string, error) {
	return CanonicalizeJSON(copyWithoutKey(capability, "signature"))
}

func VerifyCapability(capability map[string]any, now int64) (CapabilityVerification, error) {
	issuedAt, err := int64Field(capability, "issued_at")
	if err != nil {
		return CapabilityVerification{}, err
	}
	expiresAt, err := int64Field(capability, "expires_at")
	if err != nil {
		return CapabilityVerification{}, err
	}
	timeStatus := CapabilityTimeStatusValid
	switch {
	case now < issuedAt:
		timeStatus = CapabilityTimeStatusNotYetValid
	case now >= expiresAt:
		timeStatus = CapabilityTimeStatusExpired
	}

	body, err := CapabilityBodyCanonicalJSON(capability)
	if err != nil {
		return CapabilityVerification{}, err
	}
	issuer, err := stringField(capability, "issuer")
	if err != nil {
		return CapabilityVerification{}, err
	}
	signature, err := stringField(capability, "signature")
	if err != nil {
		return CapabilityVerification{}, err
	}
	signatureValid, err := VerifyUTF8MessageEd25519(body, issuer, signature)
	if err != nil {
		return CapabilityVerification{}, err
	}
	delegationChainValid, err := verifyDelegationChain(capability["delegation_chain"])
	if err != nil {
		return CapabilityVerification{}, err
	}

	return CapabilityVerification{
		DelegationChainValid: delegationChainValid,
		SignatureValid:       signatureValid,
		TimeStatus:           timeStatus,
		TimeValid:            timeStatus == CapabilityTimeStatusValid,
	}, nil
}

func VerifyCapabilityJSON(input string, now int64) (CapabilityVerification, error) {
	capability, err := ParseCapabilityJSON(input)
	if err != nil {
		return CapabilityVerification{}, err
	}
	return VerifyCapability(capability, now)
}

func verifyDelegationChain(raw any) (bool, error) {
	if raw == nil {
		return true, nil
	}
	entries, ok := raw.([]any)
	if !ok {
		return false, newInvariantError("json", "delegation_chain must be an array")
	}
	var previous map[string]any
	for _, entry := range entries {
		current, ok := entry.(map[string]any)
		if !ok {
			return false, newInvariantError("json", "delegation_chain entries must be objects")
		}
		body, err := CanonicalizeJSON(copyWithoutKey(current, "signature"))
		if err != nil {
			return false, err
		}
		delegator, err := stringField(current, "delegator")
		if err != nil {
			return false, err
		}
		signature, err := stringField(current, "signature")
		if err != nil {
			return false, err
		}
		signatureValid, err := VerifyUTF8MessageEd25519(body, delegator, signature)
		if err != nil {
			return false, err
		}
		if !signatureValid {
			return false, nil
		}
		if previous != nil {
			previousDelegatee, err := stringField(previous, "delegatee")
			if err != nil {
				return false, err
			}
			if previousDelegatee != delegator {
				return false, nil
			}
			previousTimestamp, err := int64Field(previous, "timestamp")
			if err != nil {
				return false, err
			}
			currentTimestamp, err := int64Field(current, "timestamp")
			if err != nil {
				return false, err
			}
			if currentTimestamp < previousTimestamp {
				return false, nil
			}
		}
		previous = current
	}
	return true, nil
}
